### PR TITLE
feat: Validate Git tokens before consuming them

### PR DESCRIPTION
**Pull request by kylecarbs**
This works the exact same way that the Git credential manager does. It ensures the user token is valid before returning it to the client.

It's been manually tested on GitHub, GitLab, and BitBucket.

### CODE REVIEW

**Comment by kylecarbs** [comment]
@ericpaulsen when you get a shot, we should make some docs that explain how to set Git authentication up for providers like BitBucket, GitHub Enterprise, and GitLab self-hosted. This will involve setting the following values:

```yaml
# GitHub Enterprise
CODER_GITAUTH_0_VALIDATE_URL=https://<hostname>/api/v3/user
CODER_GITAUTH_0_AUTH_URL=https://<hostname>/login/oauth/authorize
CODER_GITAUTH_0_TOKEN_URL=https://<hostname>/login/oauth/access_token
```

We should do this for the other providers and manually test that it works as well when adding docs. The validate URL is added in this PR and is required.

File: coderd/workspaceagents.go [hunk 1]
```diff
@@ -1213,6 +1220,24 @@ func (api *API) workspaceAgentsGitAuth(rw http.ResponseWriter, r *http.Request)
 		return
 	}
 
+	if gitAuthConfig.ValidateURL != "" {
+		valid, err := validateGitToken(r.Context(), gitAuthConfig.ValidateURL, token.AccessToken)
+		if err != nil {
+			httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
```
Comment (mafredri) [inline coderd/workspaceagents.go:1226]:
```suggestion
		valid, err := validateGitToken(ctx, gitAuthConfig.ValidateURL, token.AccessToken)
		if err != nil {
			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
```

File: coderd/workspaceagents.go [hunk 2]
```diff
@@ -1234,6 +1259,30 @@ func (api *API) workspaceAgentsGitAuth(rw http.ResponseWriter, r *http.Request)
...
+		return false, err
+	}
+	defer res.Body.Close()
+	if res.StatusCode == http.StatusUnauthorized {
+		// The token is no longer valid!
+		return false, nil
+	}
+	if res.StatusCode != http.StatusOK {
+		data, _ := io.ReadAll(res.Body)
+		return false, xerrors.Errorf("body: %s", data)
```
Comment (mafredri) [inline coderd/workspaceagents.go:1281]: Suggestion: More information will help us track issues down when some provider returns e.g. http no content or whatever they can think up. 😄

```suggestion
		return false, xerrors.Errorf("git token validation failed: status %d: body: %s", res.StatusCode, data)
```

File: coderd/workspaceagents_test.go [hunk 3]
```diff
@@ -934,6 +935,74 @@ func TestWorkspaceAgentsGitAuth(t *testing.T) {
...
+
+		resp := gitAuthCallback(t, "github", client)
+		require.Equal(t, http.StatusTemporaryRedirect, resp.StatusCode)
+
+		// If the validation URL says unauthorized, the callback
+		// URL to re-authenticate should be returned.
+		srv.Config.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusUnauthorized)
+		})
+		res, err := agentClient.WorkspaceAgentGitAuth(context.Background(), "github.com/asd/asd", false)
```
Comment (mafredri) [inline coderd/workspaceagents_test.go:990]: Suggestion: Consider using `testutil.Context` or similar timeout mechanism here and below.

File: coderd/workspaceagents.go [hunk 4]
```diff
@@ -1158,6 +1159,12 @@ func (api *API) workspaceAgentsGitAuth(rw http.ResponseWriter, r *http.Request)
 			if gitAuthLink.OAuthExpiry.Before(database.Now()) {
 				continue
 			}
+			if gitAuthConfig.ValidateURL != "" {
+				valid, _ := validateGitToken(ctx, gitAuthConfig.ValidateURL, gitAuthLink.OAuthAccessToken)
```
Comment (mafredri) [inline coderd/workspaceagents.go:1163]: Nit: We should log this errors, otherwise we won't know why it's not working.

**Review by mafredri — COMMENTED** [review]
Some minor stuff, looks good in general. 👍🏻

**Review by deansheather — APPROVED** [review]
LGTM other than Mathias' comments

**Pull request merged**